rfcs/control-groups/README.md

```diff
@@ -0,0 +1,279 @@
+and have implementation consistency.
+
+First thing that comes to mind in this space is [Wiki: Linux Control Groups](https://en.wikipedia.org/wiki/Cgroups)
+([kernel.org: cgroup-v2.txt](https://www.kernel.org/doc/Documentation/cgroup-v2.txt)).
+
+> cgroup is a mechanism to organize processes hierarchically and distribute
+> system resources along the hierarchy in a controlled and configurable manner.
+> cgroup is largely composed of two parts - the core and controllers.
+>
+> cgroup core is primarily responsible for hierarchically organizing processes.
+>
+> A cgroup controller is usually responsible for distributing a specific type of
+> system resource along the hierarchy although there are utility controllers
+> which serve purposes other than resource distribution.
+>
+> cgroups form a tree structure and every process in the system belongs to
+> one and only one cgroup.
+
+![cgroupv2 hierarchy](./cgroupv2-hierarchy.jpg)
+
+
+### What we have
+
+ClickHouse already has things similar to cgroup controllers that are used to
+implement various limits.
+
+Examples: [MemoryTracker][MemoryTracker.h], Network [Throttler][Throttler.h].
+
+[ProfileEvents][ProfileEvents.h] and thread groups hierarchies with
+[ThreadProfileEvents][ThreadProfileEvents.h] could fall in this category as well
+even though today only tracking is supported (rather than "controlling").
+[3] has some neat ideas on how it could be possible to "bind" actions to events.
+
+An important property to mention of the above mentioned "things" is that all of
+them are hierarchical.
+
+The missing part is that there isn't a lot of flexibility in configuring these.
+
+
+### A rough sketch of how the future could look like
+
+Let's draw some inspiration from Linux control groups and adapt the existing
+"resource controllers" to achieve "resource pools".
+
+* There should be a root control group for the system limits
+
+* It should be possible to define control groups for background processes
+and assign tables to these control groups.
+    * Ie a table or a group of tables could use at most `X b/s` of disk, 
+    `Y b/s` of network for replication, `Z b` of storage.
+        * We could instead use writer's control group but there are things like
+        Kafka engine which would need some special handling.
+
+* It should be possible to define control groups for queries to limit eg CPU,
+query concurrency, max memory usage.
+    * This could help isolate writing vs reading.
+
+* It should be possible to assign users/roles to control groups.
+    * Maybe even multiple ones and allow users to choose which control groups
+    should be used for operations (similar to roles today?).
+
+* It should be possible to nest control groups to form hierarchies.
+
+* It should be possible to configure per-user and per-query limits using control
+groups in order to achieve "consistent implementation" of the limits.
+    * Maybe allow something like `CREATE CONTROL GROUP TEMPLATE zzz FOR QUERY ...`,
```

Comment:
This idea is not clear. Probably it is too complex without prepared queries to set some settings for some query type